File: backend/models/refund_request.py
from __future__ import annotations

from typing import Optional, List, Literal, Dict, Any
from datetime import datetime, timezone

from bson import ObjectId
from pydantic import BaseModel, Field

from mongo.database import DB


# -----------------------------
# Models
# -----------------------------

RefundTxnKind = Literal["event", "form"]


class RefundRequestItem(BaseModel):
    message: Optional[str] = None
    responded: bool
    resolved: bool
    reason: Optional[str] = None
    created_on: datetime
    responded_to: Optional[datetime] = None


class RefundRequest(BaseModel):
    uid: str
    txn_kind: RefundTxnKind
    txn_id: str

    # User-provided message (required at API layer)
    message: str

    # Admin state
    responded: bool = False
    resolved: bool = False
    reason: Optional[str] = None

    created_on: datetime = Field(default_factory=datetime.now)
    responded_to: Optional[datetime] = None

    history: List[RefundRequestItem] = Field(default_factory=list)


class RefundRequestOut(RefundRequest):
    # MongoDB _id as string, plus optional transaction payload
    id: str
    transaction: Optional[Any] = None


class RefundRequestIn(BaseModel):
    txn_kind: RefundTxnKind
    txn_id: str
    message: str


class RefundRequestAdminUpdate(BaseModel):
    id: str
    responded: bool
    resolved: bool
    reason: Optional[str] = None


class RefundRequestSearchParams(BaseModel):
    page: int
    pageSize: int
    status: Literal["pending", "resolved", "unresolved", "all"] = "pending"
    txn_kind: Optional[RefundTxnKind] = None
    uid: Optional[str] = None  # optional filter; user routes will set this


# -----------------------------
# Internal helpers
# -----------------------------


def _status_match_for_params(params: RefundRequestSearchParams) -> Dict[str, Any]:
    if params.status == "pending":
        return {"responded": False}
    if params.status == "resolved":
        return {"responded": True, "resolved": True}
    if params.status == "unresolved":
        return {"responded": True, "resolved": False}
    return {}  # "all"


def _to_out_from_doc(doc: Dict[str, Any]) -> RefundRequestOut:
    history_raw = doc.get("history", []) or []
    history_items: List[RefundRequestItem] = [
        RefundRequestItem(**h) for h in history_raw
    ]

    return RefundRequestOut(
        id=str(doc.get("_id")),
        uid=doc["uid"],
        txn_kind=doc["txn_kind"],
        txn_id=str(doc.get("txn_id")),
        message=doc.get("message", ""),
        responded=doc.get("responded", False),
        resolved=doc.get("resolved", False),
        reason=doc.get("reason"),
        created_on=doc.get("created_on"),
        responded_to=doc.get("responded_to"),
        history=history_items,
        transaction=None,
    )


# -----------------------------
# CRUD helpers
# -----------------------------
# (no cross-collection logic here; controller handles that)
# -----------------------------

# CREATE / UPDATE (user-initiated)


async def create_or_update_refund_request_doc(uid: str, payload: RefundRequestIn) -> Dict[str, Any]:
    """
    Low-level create/update operation for the refund_requests collection.
    Does not enforce transaction ownership/eligibility – that belongs in the controller.
    """
    coll = DB.db["refund_requests"]

    message = (payload.message or "").strip()

    existing = await coll.find_one(
        {
            "uid": uid,
            "txn_kind": payload.txn_kind,
            "txn_id": payload.txn_id,
        }
    )

    if not existing:
        new_doc = RefundRequest(
            uid=uid,
            txn_kind=payload.txn_kind,
            txn_id=payload.txn_id,
            message=message,
        ).model_dump()
        res = await coll.insert_one(new_doc)
        ok = getattr(res, "inserted_id", None) is not None
        return {
            "success": bool(ok),
            "msg": "Created refund request" if ok else "Failed to create refund request",
        }

    history_item = RefundRequestItem(
        message=existing.get("message"),
        responded=existing.get("responded", False),
        resolved=existing.get("resolved", False),
        reason=existing.get("reason"),
        created_on=existing.get("created_on", datetime.now(timezone.utc)),
        responded_to=existing.get("responded_to"),
    ).model_dump()

    update = {
        "$push": {"history": history_item},
        "$set": {
            "message": message,
            "responded": False,
            "resolved": False,
            "reason": None,
            "created_on": datetime.now(timezone.utc),
            "responded_to": None,
        },
    }

    res = await coll.update_one({"_id": existing["_id"]}, update)
    ok = res.matched_count > 0
    return {
        "success": bool(ok),
        "msg": "Updated refund request with new entry" if ok else "Failed to update refund request",
    }


# READ – by id / by user + transaction


async def get_refund_request_by_id(id: str) -> Optional[Dict[str, Any]]:
    coll = DB.db["refund_requests"]
    oid = ObjectId(id) if ObjectId.is_valid(id) else id
    doc = await coll.find_one({"_id": oid})
    return doc


async def get_refund_request_by_user_and_txn(
    uid: str,
    txn_kind: RefundTxnKind,
    txn_id: str,
) -> Optional[Dict[str, Any]]:
    coll = DB.db["refund_requests"]
    doc = await coll.find_one(
        {"uid": uid, "txn_kind": txn_kind, "txn_id": txn_id}
    )
    return doc


# READ – search (paged) with filters


async def search_refund_requests(params: RefundRequestSearchParams) -> Dict[str, Any]:
    """
    Server-side search/paging/sort for refund requests.

    Filters:
      - uid (optional)
      - txn_kind (event / form)
      - status (pending / resolved / unresolved / all)
    """
    coll = DB.db["refund_requests"]

    match: Dict[str, Any] = {}

    if params.uid:
        match["uid"] = params.uid
    if params.txn_kind:
        match["txn_kind"] = params.txn_kind

    status_match = _status_match_for_params(params)
    if status_match:
        match.update(status_match)

    pipeline: List[Dict[str, Any]] = [
        {"$match": match},
        {"$sort": {"created_on": -1}},
        {
            "$facet": {
                "total": [{"$count": "count"}],
                "items": [
                    {
                        "$skip": max(0, params.page) * max(1, params.pageSize)
                    },
                    {"$limit": max(1, params.pageSize)},
                ],
            }
        },
    ]

    agg = await coll.aggregate(pipeline).to_list(length=1)
    bucket = agg[0] if agg else {"items": [], "total": []}

    total_arr = bucket.get("total") or []
    total = total_arr[0].get("count", 0) if total_arr else 0

    items_raw = bucket.get("items", []) or []
    items: List[RefundRequestOut] = [
        _to_out_from_doc(doc) for doc in items_raw
    ]

    return {
        "items": [i.model_dump() for i in items],
        "total": total,
        "page": params.page,
        "pageSize": params.pageSize,
    }


# UPDATE – admin POV


async def update_refund_request_admin(update: RefundRequestAdminUpdate) -> Dict[str, Any]:
    coll = DB.db["refund_requests"]
    oid = ObjectId(update.id) if ObjectId.is_valid(update.id) else update.id

    existing = await coll.find_one({"_id": oid})
    if not existing:
        return {"success": False, "msg": "Refund request not found"}

    history_item = RefundRequestItem(
        message=existing.get("message"),
        responded=existing.get("responded", False),
        resolved=existing.get("resolved", False),
        reason=existing.get("reason"),
        created_on=existing.get("created_on", datetime.now(timezone.utc)),
        responded_to=existing.get("responded_to"),
    ).model_dump()

    updates: Dict[str, Any] = {
        "responded": bool(update.responded),
        "resolved": bool(update.resolved),
        "responded_to": datetime.now(timezone.utc),
    }

    if update.reason is not None:
        updates["reason"] = update.reason.strip()

    res = await coll.update_one(
        {"_id": oid},
        {
            "$push": {"history": history_item},
            "$set": updates,
        },
    )

    if res.matched_count == 0:
        return {"success": False, "msg": "Refund request not found"}

    return {"success": True, "msg": "Refund request updated"}


# UPDATE – explicit low-level update helper


async def explicit_update_refund_request(
    filter_query: Dict[str, Any],
    update_data: Dict[str, Any],
):
    return await DB.update_document("refund_requests", filter_query, update_data)


# DELETE – by id


async def delete_refund_request(id: str) -> Dict[str, Any]:
    coll = DB.db["refund_requests"]
    oid = ObjectId(id) if ObjectId.is_valid(id) else id
    res = await coll.delete_one({"_id": oid})
    if res.deleted_count == 0:
        return {"success": False, "msg": "Refund request not found"}
    return {"success": True, "msg": "Refund request deleted"}

File: frontend/app/lib/models/refund_request.dart
// refund_request.dart
// Dart models for refund requests (user + shared core; no admin).

import 'transactions.dart';

enum RefundTxnKind { event, form }

RefundTxnKind? refundTxnKindFromJson(String? value) {
  switch (value) {
    case 'event':
      return RefundTxnKind.event;
    case 'form':
      return RefundTxnKind.form;
    default:
      return null;
  }
}

String? refundTxnKindToJson(RefundTxnKind? value) {
  if (value == null) return null;
  switch (value) {
    case RefundTxnKind.event:
      return 'event';
    case RefundTxnKind.form:
      return 'form';
  }
}

enum RefundRequestStatus { pending, resolved, unresolved, all }

RefundRequestStatus? refundRequestStatusFromJson(String? value) {
  switch (value) {
    case 'pending':
      return RefundRequestStatus.pending;
    case 'resolved':
      return RefundRequestStatus.resolved;
    case 'unresolved':
      return RefundRequestStatus.unresolved;
    case 'all':
      return RefundRequestStatus.all;
    default:
      return null;
  }
}

String? refundRequestStatusToJson(RefundRequestStatus? value) {
  if (value == null) return null;
  switch (value) {
    case RefundRequestStatus.pending:
      return 'pending';
    case RefundRequestStatus.resolved:
      return 'resolved';
    case RefundRequestStatus.unresolved:
      return 'unresolved';
    case RefundRequestStatus.all:
      return 'all';
  }
}

class RefundRequestHistoryItem {
  final String? message;
  final bool responded;
  final bool resolved;
  final String? reason;
  final String createdOn; // ISO timestamp
  final String? respondedTo; // ISO timestamp

  RefundRequestHistoryItem({
    this.message,
    required this.responded,
    required this.resolved,
    this.reason,
    required this.createdOn,
    this.respondedTo,
  });

  factory RefundRequestHistoryItem.fromJson(Map<String, dynamic> json) {
    return RefundRequestHistoryItem(
      message: json['message'] as String?,
      responded: json['responded'] as bool? ?? false,
      resolved: json['resolved'] as bool? ?? false,
      reason: json['reason'] as String?,
      createdOn: json['created_on'] as String? ?? '',
      respondedTo: json['responded_to'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'message': message,
      'responded': responded,
      'resolved': resolved,
      'reason': reason,
      'created_on': createdOn,
      'responded_to': respondedTo,
    };
  }
}

class RefundRequest {
  final String uid;
  final RefundTxnKind txnKind;
  final String txnId;

  final String message;

  final bool responded;
  final bool resolved;
  final String? reason;

  final String createdOn; // ISO timestamp
  final String? respondedTo; // ISO timestamp

  final List<RefundRequestHistoryItem> history;

  RefundRequest({
    required this.uid,
    required this.txnKind,
    required this.txnId,
    required this.message,
    required this.responded,
    required this.resolved,
    this.reason,
    required this.createdOn,
    this.respondedTo,
    required this.history,
  });

  factory RefundRequest.fromJson(Map<String, dynamic> json) {
    return RefundRequest(
      uid: json['uid'] as String? ?? '',
      txnKind:
          refundTxnKindFromJson(json['txn_kind'] as String?) ??
          RefundTxnKind.event,
      txnId: json['txn_id'] as String? ?? '',
      message: json['message'] as String? ?? '',
      responded: json['responded'] as bool? ?? false,
      resolved: json['resolved'] as bool? ?? false,
      reason: json['reason'] as String?,
      createdOn: json['created_on'] as String? ?? '',
      respondedTo: json['responded_to'] as String?,
      history:
          (json['history'] as List<dynamic>? ?? [])
              .map(
                (e) => RefundRequestHistoryItem.fromJson(
                  Map<String, dynamic>.from(e as Map),
                ),
              )
              .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'uid': uid,
      'txn_kind': refundTxnKindToJson(txnKind),
      'txn_id': txnId,
      'message': message,
      'responded': responded,
      'resolved': resolved,
      'reason': reason,
      'created_on': createdOn,
      'responded_to': respondedTo,
      'history': history.map((e) => e.toJson()).toList(),
    };
  }
}

// RefundRequest with attached transaction summary (as returned by the
// search endpoints).
class RefundRequestWithTransaction extends RefundRequest {
  final String id; // Mongo _id
  final TransactionSummary? transaction; // unified transaction payload

  RefundRequestWithTransaction({
    required this.id,
    required this.transaction,
    required String uid,
    required RefundTxnKind txnKind,
    required String txnId,
    required String message,
    required bool responded,
    required bool resolved,
    String? reason,
    required String createdOn,
    String? respondedTo,
    required List<RefundRequestHistoryItem> history,
  }) : super(
         uid: uid,
         txnKind: txnKind,
         txnId: txnId,
         message: message,
         responded: responded,
         resolved: resolved,
         reason: reason,
         createdOn: createdOn,
         respondedTo: respondedTo,
         history: history,
       );

  factory RefundRequestWithTransaction.fromJson(Map<String, dynamic> json) {
    final base = RefundRequest.fromJson(json);
    return RefundRequestWithTransaction(
      id: json['id'] as String? ?? '',
      transaction:
          json['transaction'] != null
              ? TransactionSummary.fromJson(
                Map<String, dynamic>.from(json['transaction'] as Map),
              )
              : null,
      uid: base.uid,
      txnKind: base.txnKind,
      txnId: base.txnId,
      message: base.message,
      responded: base.responded,
      resolved: base.resolved,
      reason: base.reason,
      createdOn: base.createdOn,
      respondedTo: base.respondedTo,
      history: base.history,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    final data = super.toJson();
    data['id'] = id;
    data['transaction'] = transaction?.toJson();
    return data;
  }
}

// Payload for creating a refund request.
class CreateRefundRequestPayload {
  final RefundTxnKind txnKind;
  final String txnId;
  final String message;

  CreateRefundRequestPayload({
    required this.txnKind,
    required this.txnId,
    required this.message,
  });

  factory CreateRefundRequestPayload.fromJson(Map<String, dynamic> json) {
    return CreateRefundRequestPayload(
      txnKind:
          refundTxnKindFromJson(json['txn_kind'] as String?) ??
          RefundTxnKind.event,
      txnId: json['txn_id'] as String? ?? '',
      message: json['message'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'txn_kind': refundTxnKindToJson(txnKind),
      'txn_id': txnId,
      'message': message,
    };
  }
}

// Simple success envelope for create.
class CreateRefundRequestResponse {
  final bool success;
  final String? msg;

  CreateRefundRequestResponse({required this.success, this.msg});

  factory CreateRefundRequestResponse.fromJson(Map<String, dynamic> json) {
    return CreateRefundRequestResponse(
      success: json['success'] as bool? ?? false,
      msg: json['msg'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{'success': success, 'msg': msg};
  }
}

// Query params/body for "my refund requests" listing.
class MyRefundRequestSearchParams {
  final int? page; // 0-based
  final int? pageSize;
  final RefundRequestStatus? status;
  final RefundTxnKind? txnKind;

  MyRefundRequestSearchParams({
    this.page,
    this.pageSize,
    this.status,
    this.txnKind,
  });

  factory MyRefundRequestSearchParams.fromJson(Map<String, dynamic> json) {
    return MyRefundRequestSearchParams(
      page: (json['page'] as num?)?.toInt(),
      pageSize: (json['pageSize'] as num?)?.toInt(),
      status: refundRequestStatusFromJson(json['status'] as String?),
      txnKind: refundTxnKindFromJson(json['txn_kind'] as String?),
    );
  }

  Map<String, dynamic> toJson() {
    final data = <String, dynamic>{};
    if (page != null) data['page'] = page;
    if (pageSize != null) data['pageSize'] = pageSize;
    final statusStr = refundRequestStatusToJson(status);
    if (statusStr != null) data['status'] = statusStr;
    final kindStr = refundTxnKindToJson(txnKind);
    if (kindStr != null) data['txn_kind'] = kindStr;
    return data;
  }
}

// Paged results envelope for refund requests.
class RefundRequestPagedResults {
  final List<RefundRequestWithTransaction> items;
  final int total;
  final int page;
  final int pageSize;

  RefundRequestPagedResults({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  factory RefundRequestPagedResults.fromJson(Map<String, dynamic> json) {
    return RefundRequestPagedResults(
      items:
          (json['items'] as List<dynamic>? ?? [])
              .map(
                (e) => RefundRequestWithTransaction.fromJson(
                  Map<String, dynamic>.from(e as Map),
                ),
              )
              .toList(),
      total: (json['total'] as num?)?.toInt() ?? 0,
      page: (json['page'] as num?)?.toInt() ?? 0,
      pageSize: (json['pageSize'] as num?)?.toInt() ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'items': items.map((e) => e.toJson()).toList(),
      'total': total,
      'page': page,
      'pageSize': pageSize,
    };
  }
}
